homeassistant/components/apple_tv/__init__.py

```diff
@@ -70,205 +88,252 @@ def ensure_list(value: Union[T, Sequence[T]]) -> Sequence[T]:
+        """Device connection was (intentionally) closed."""
+        self.atv = None
+        self._update_state(disconnected=True)
+        self._start_connect_loop()
+
+    async def connect(self):
+        """Connect to device."""
+        self._is_on = True
+        self._start_connect_loop()
+
+    async def disconnect(self):
+        """Disconnect from device."""
+        _LOGGER.debug("Disconnecting from device")
+        self._is_on = False
+        try:
+            if self.atv:
+                self.atv.push_updater.listener = None
+                self.atv.push_updater.stop()
+                await self.atv.close()
+                self.atv = None
+            if self._task:
+                self._task.cancel()
+                self._task = None
+        finally:
+            self._update_state(disconnected=False)
+
+    def _start_connect_loop(self):
+        if not self._task and self.atv is None and self._is_on:
+            self._task = asyncio.ensure_future(
+                self._connect_loop(), loop=self.hass.loop
+            )
         else:
-            devices = hass.data[DATA_ENTITIES]
-
-        for device in devices:
-            if service.service != SERVICE_AUTHENTICATE:
-                continue
-
-            atv = device.atv
-            credentials = await atv.airplay.generate_credentials()
-            await atv.airplay.load_credentials(credentials)
-            _LOGGER.debug("Generated new credentials: %s", credentials)
-            await atv.airplay.start_authentication()
-            hass.async_add_job(request_configuration, hass, config, atv, credentials)
-
-    async def atv_discovered(service, info):
-        """Set up an Apple TV that was auto discovered."""
-        await _setup_atv(
-            hass,
-            config,
-            {
-                CONF_NAME: info["name"],
-                CONF_HOST: info["host"],
-                CONF_LOGIN_ID: info["properties"]["hG"],
-                CONF_START_OFF: False,
-            },
-        )
-
-    discovery.async_listen(hass, SERVICE_APPLE_TV, atv_discovered)
+            _LOGGER.debug(
+                "Not starting connect loop (%s, %s)", self.atv is None, self._is_on
+            )
 
-    tasks = [_setup_atv(hass, config, conf) for conf in config.get(DOMAIN, [])]
-    if tasks:
-        await asyncio.wait(tasks)
+    async def _connect_loop(self):
+        _LOGGER.debug("Starting connect loop")
+
+        # Try to find device and connect as long as the user has said that
+        # we are allowed to connect and we are not already connected.
+        while self._is_on and self.atv is None:
+            try:
+                conf = await self._scan()
+                if conf:
+                    await self._connect(conf)
+            except exceptions.AuthenticationError:
+                self._auth_problem()
+                break
+            except asyncio.CancelledError:
+                pass
+            except Exception:
```

Comment:
Catching broad exception, please only catch the exceptions that could occur. 
Or if this cannot be avoided add `# pylint: disable=broad-except` to solve the pylint error.

Reply:
I have added exclusion locally. This is one of those places where I *really* want to catch everything since the re-connection logic breaks otherwise. It's too many things that can go wrong so it would be so much unnecessary work to list them all (and also hard to find).

Reply:
If that is the case I would suggest to log the exception so it is easier to understand what went wrong for you and the user.

Reply:
It is logged on the next line with _LOGGER.exception.